src/GafferScene/RendererAlgo.cpp

```diff
@@ -1542,6 +1589,27 @@ void applyCameraGlobals( IECoreScene::Camera *camera, const IECore::CompoundObje
 
 }
 
+ConstShaderNetworkPtr substitutedShader( IECoreScene::ConstShaderNetworkPtr shaderNetwork, const IECore::CompoundObject *attributes )
+{
+	// Fixed size cache that will hold 10000 shaders
+	static SubstitutionCache substitutionCache( substitutionCacheGetter, 10000 );
```

Comment:
The reason why I hesitated to use a memory limit here is that my understanding was that the memory limit wouldn't be particularly accurate.  memoryUsage() has no understanding of sharing, right?  So it would dramatically overestimate the memory used by assuming that all shaders are copied, rather than just the ones that are substituted?

It would be an upper bound, but it feels like such an inaccurate upper bound as to be almost meaningless?

I guess I could change the API on applySubstitutions to return the amount of additional memory used by the shaders it actually needed to copy?  That doesn't cover the case where the original shader gets released, and the substituted one is left holding the references, but it feels like a more accurate heuristic for memory used than recounting every shader in the network?

Reply:
> The reason why I hesitated to use a memory limit here is that my understanding was that the memory limit wouldn't be particularly accurate. memoryUsage() has no understanding of sharing, right? So it would dramatically overestimate the memory used by assuming that all shaders are copied, rather than just the ones that are substituted?

Hmm, true. Good point. OK, let's stick with the number-of-shaders limit, but do a bit of due diligence to check that we have the right constant.